### PR TITLE
GitHub CI/CD Workflow: Refactor dependencies and git management

### PR DESCRIPTION
**Pull request by jcortejoso**
### Description

Refactor the GitHub CI/CD Workflow to handle recent changes and bugs on some jobs:
- Fetch git submodules
- Removed not needed steps for GitHub Actions
- Removed circleci.yaml
- Removed legacy scripts used by CircleCI CI
- Update action/checkout to v4
- Use GitHub artifacts for generating protocol devchain with a new workflow (it runs monthly and also can be executed from `Actions` menu)

### CODE REVIEW

**Comment by aaronmgdr** [comment]
A HERO!

File: .github/workflows/celo-monorepo.yml [hunk 1]
```diff
@@ -358,35 +251,24 @@ jobs:
...
-          key: git-${{ github.ref }}
-      - uses: actions/checkout@v3
+      - uses: actions/checkout@v4
         with:
           submodules: recursive
       - name: Sync workspace
         uses: ./.github/actions/sync-workspace
         with:
-          package-json-checksum: ${{ needs.install-dependencies.outputs.package-json-checksum }}
           rebuild-package: 'true'
```
Comment (mcortesi) [inline .github/workflows/celo-monorepo.yml:260]: do we need to rebuild here? 
We start with dependencies and compile artifacts from the restore/checkout, so doesn't seem like it

**Review by mcortesi — APPROVED** [review]

**Comment by aaronmgdr** [comment]
if script to update dependency group is being removed id think the dependency-graph.json file itself would be too

File: .github/workflows/protocol-devchain.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+jobs:
+  generate-protocol-devchain:
+    name: Generate protocol devchain used in celo-monorepo.yml workflow
+    runs-on: ['self-hosted', 'monorepo-node18']
+    timeout-minutes: 60
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - tag: core-contracts.v9
```
Comment (soloseng) [inline .github/workflows/protocol-devchain.yml:19]: Not sure the node12 compatibility is still needed.

**Review by soloseng — APPROVED** [review]

File: .github/workflows/celo-monorepo.yml (hunk 1 above)
Comment (mcortesi) [reply .github/workflows/celo-monorepo.yml:260]: checkon other branch and it seems it's needed

File: .github/workflows/protocol-devchain.yml (hunk 2 above)
Comment (mcortesi) [reply .github/workflows/protocol-devchain.yml:19]: we can remove it later

**Pull request merged**